{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "022bf988",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests  \n",
    "import re\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from Bio import Medline\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from Bio import SearchIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "import mysql.connector as SQLC\n",
    "import mysql.connector\n",
    "from datetime import date\n",
    "import time\n",
    "from bs4 import BeautifulSoup\n",
    "from Bio.SeqFeature import CompoundLocation\n",
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "e05384b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "escolha o que quer pesquisar: virus\n",
      "escolha o nº de genes que quer obter (máximo 20): 5\n",
      "['https://www.ncbi.nlm.nih.gov/nuccore/V01302.1', 'https://www.ncbi.nlm.nih.gov/nuccore/X67705.1']\n"
     ]
    }
   ],
   "source": [
    "#Extrair ids genebank:\n",
    "data_e_hora_atuais = datetime.now()\n",
    "\n",
    "query= input('escolha o que quer pesquisar: ')\n",
    "\n",
    "def url_get(i):\n",
    "    url_list= [ ]\n",
    "    url = \"https://www.ncbi.nlm.nih.gov/gene/?term={}\".format( query )\n",
    "    url_list.append(url)\n",
    "    return url_list\n",
    "url_get(1)\n",
    "\n",
    "content = []\n",
    "for url in url_get(1):\n",
    "    r = requests.get(url)\n",
    "    content.append(r.content)\n",
    "    \n",
    "for c in content:\n",
    "    soup = BeautifulSoup(c, 'html.parser')\n",
    "    a= soup.get_text()\n",
    "    \n",
    "\n",
    "existe = re.findall(r\"ID:\\s+\\d*(?=\\D)\", a, re.DOTALL)\n",
    "\n",
    "c= ', '.join(existe)\n",
    "h= c.replace('ID: ','')\n",
    "IDS= h.split(', ')\n",
    "\n",
    "n_genes= IDS[0:int(input('escolha o nº de genes que quer obter (máximo 20): '))]\n",
    "numero_genes= len(n_genes)\n",
    "\n",
    "#Extrair ids ncbi:\n",
    "\n",
    "Ids=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "for info in records:\n",
    "    Ids.append(info.id)\n",
    "\n",
    "#extrair links ncbi\n",
    "def url_get_id(i):\n",
    "    url_list= [ ]\n",
    "    for id in i:\n",
    "        url = \"https://www.ncbi.nlm.nih.gov/nuccore/{}\".format( id )\n",
    "        url_list.append(url)\n",
    "    return url_list\n",
    "link_genebank= url_get_id(Ids)\n",
    "print(link_genebank)\n",
    "\n",
    "#Extrair description NCBI:\n",
    "description=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "for info in records:\n",
    "    description.append(info.description)\n",
    "\n",
    "#organismos\n",
    "\n",
    "Organismos=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "for info in records:\n",
    "    Organismos.append(info.annotations['organism'])\n",
    "    \n",
    "#SEQ\n",
    "seq=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "for info in records:\n",
    "    seq.append(f'{info.seq[0:10]}...{info.seq[-10:]}')\n",
    "\n",
    "#percentagem de nucle:\n",
    "Adenina=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "seq=[]\n",
    "for info in records:\n",
    "    c=info.seq\n",
    "    count= c.count('A')\n",
    "    perc= int(count/len(c)*100)\n",
    "    Adenina.append(perc)\n",
    "\n",
    "Citosina=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "seq=[]\n",
    "for info in records:\n",
    "    c=info.seq\n",
    "    count= c.count('C')\n",
    "    perc= int(count/len(c)*100)\n",
    "    Citosina.append(perc)\n",
    "\n",
    "Guanina=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "seq=[]\n",
    "for info in records:\n",
    "    c=info.seq\n",
    "    count= c.count('G')\n",
    "    perc= int(count/len(c)*100)\n",
    "    Guanina.append(perc)\n",
    "\n",
    "Timina=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "seq=[]\n",
    "for info in records:\n",
    "    c=info.seq\n",
    "    count= c.count('T')\n",
    "    perc= int(count/len(c)*100)\n",
    "    Timina.append(perc)\n",
    "    \n",
    "#data\n",
    "data=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "for info in records:\n",
    "    data.append(info.annotations['date'])\n",
    "\n",
    "#len(SEQ)\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "tam=[]\n",
    "for info in records:\n",
    "    tam.append(len(info.seq))\n",
    "\n",
    "# buscar info pubmed\n",
    "content_id = []\n",
    "for url in url_get_id(Ids):\n",
    "    r_id = requests.get(url)\n",
    "    content_id.append(r_id.content)\n",
    "    \n",
    "from bs4 import BeautifulSoup\n",
    "listas=[]\n",
    "\n",
    "for c in content_id:\n",
    "    soup_id = BeautifulSoup(c, 'html.parser')\n",
    "\n",
    "    lines = soup_id.find_all('meta', {'name':\"ncbi_uidlist\"} )\n",
    "\n",
    "    id = \"\"\n",
    "    url = \"\"\n",
    "    for line in lines:\n",
    "        if 'content' in line.attrs:\n",
    "            id = line.attrs['content']\n",
    "    if id:\n",
    "        url = \"https://www.ncbi.nlm.nih.gov/sviewer/viewer.fcgi?id={}&db=nuccore&report=genbank&conwithfeat=on&hide-cdd=on&retmode=text&maxdownloadsize=5000000\".format(id)\n",
    "    r2 = requests.get(url)\n",
    "    r4= r2.content.decode('utf-8')\n",
    "    listas.append(r4)\n",
    "    \n",
    "cc= ', '.join(listas)\n",
    "er= cc.replace('//','')\n",
    "final= er.split(', ')\n",
    "\n",
    "#criar dicionário de ids ncbi e ids pubmed\n",
    "output_dict = {}\n",
    "for x in final:\n",
    "    version = re.search(r'VERSION\\s+(.*?)\\s', x)\n",
    "    pubmed = re.search(r'PUBMED\\s+(.*?)\\s', x)\n",
    "    if version:\n",
    "        versionf=version.group(1)\n",
    "        output_dict.setdefault(version.group(1), [])\n",
    "    if pubmed:\n",
    "        output_dict[versionf].append(pubmed.group(1))\n",
    "#     else:\n",
    "       # output_dict[version.group(1)].append(\"N/A\")\n",
    "\n",
    "#Criar Listas só com ids ncbi e ids pubmed\n",
    "id_ncbii = []\n",
    "ID_PUB = []\n",
    "for key, vals in output_dict.items():\n",
    "    if vals:\n",
    "        for val in vals:\n",
    "            id_ncbii.append(key)\n",
    "            ID_PUB.append(val)\n",
    "    else:\n",
    "        id_ncbii.append(key)\n",
    "        ID_PUB.append(\"N/A\")\n",
    "\n",
    "new_list_ = []\n",
    "seen = set()\n",
    "for item in ID_PUB:\n",
    "    if item not in seen:\n",
    "        seen.add(item)\n",
    "        new_list_.append(item)\n",
    "\n",
    "so_ids_pubmed = [val for key, vals in output_dict.items() for val in vals]\n",
    "so_ids_ncbi = list(output_dict.keys())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "949a591d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CAA24609.1', 'CAA47943.1']\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'product'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Input \u001b[1;32mIn [47]\u001b[0m, in \u001b[0;36m<cell line: 166>\u001b[1;34m()\u001b[0m\n\u001b[0;32m    164\u001b[0m     handle2\u001b[38;5;241m.\u001b[39mclose()                    \n\u001b[0;32m    165\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m cds_location_list\n\u001b[1;32m--> 166\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mget_CDS_info\u001b[49m\u001b[43m(\u001b[49m\u001b[43mIds\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[0;32m    168\u001b[0m \u001b[38;5;66;03m#FINAL result_dic = {Ids, ID_PROT} - NECESSÁRIO\u001b[39;00m\n\u001b[0;32m    169\u001b[0m result_dict \u001b[38;5;241m=\u001b[39m {}\n",
      "Input \u001b[1;32mIn [47]\u001b[0m, in \u001b[0;36mget_CDS_info\u001b[1;34m(id_list)\u001b[0m\n\u001b[0;32m    148\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m i\u001b[38;5;241m.\u001b[39mtype \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCDS\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m    149\u001b[0m     i_d \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(i\u001b[38;5;241m.\u001b[39mqualifiers[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprotein_id\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[1;32m--> 150\u001b[0m     product \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(\u001b[43mi\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mqualifiers\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mproduct\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m)\n\u001b[0;32m    151\u001b[0m     translation \u001b[38;5;241m=\u001b[39m  \u001b[38;5;28mstr\u001b[39m(i\u001b[38;5;241m.\u001b[39mqualifiers[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtranslation\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[0;32m    152\u001b[0m     cds_location\u001b[38;5;241m.\u001b[39mappend(i_d)\n",
      "\u001b[1;31mKeyError\u001b[0m: 'product'"
     ]
    }
   ],
   "source": [
    "#Extrair informação de artigos\n",
    "\n",
    "titles=[]\n",
    "authors=[]\n",
    "source=[]\n",
    "affiliation=[]\n",
    "database = 'PubMed'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= new_list_\n",
    "counter = 0\n",
    "for i in idlist:\n",
    "    if i!= \"N/A\":\n",
    "        handle = Entrez.efetch(db=database, id=i, rettype=\"medline\", retmode=\"text\") \n",
    "        records = Medline.parse(handle)\n",
    "        for info in records:\n",
    "            titles.append(info.get(\"TI\", [\"N/A\"]))\n",
    "            authors_string = info.get(\"AU\", [\"N/A\"])\n",
    "            if len(authors_string) > 5:\n",
    "                authors_h = authors_string[0:5]\n",
    "                authors.append(authors_h)\n",
    "            else:\n",
    "                authors.append(authors_string) \n",
    "            source.append(info.get(\"SO\", [\"N/A\"]))\n",
    "            affiliation_string= info.get(\"AD\", [\"N/A\"])\n",
    "            if len(affiliation_string) > 5:\n",
    "                affiliation_h = affiliation_string[0:5]\n",
    "                affiliation.append(affiliation_h)\n",
    "            else:\n",
    "                affiliation.append(affiliation_string)\n",
    "            counter += 1\n",
    "    else:\n",
    "        titles.append([\"N/A\"])\n",
    "        authors.append([\"N/A\"])\n",
    "        source.append([\"N/A\"])\n",
    "        affiliation.append([\"N/A\"])\n",
    "        counter += 1\n",
    "        \n",
    "#agrupar titles\n",
    "titles = [ [title] for title in titles]\n",
    "\n",
    "# agrupar dois\n",
    "doi_list = []\n",
    "for x in source:\n",
    "    match = re.search(\"doi: (.*)\", str(x))\n",
    "    if match:\n",
    "        doi_list.append(match.group(1))\n",
    "    else:\n",
    "        doi_list.append(\"N/A\")\n",
    "        \n",
    "#agrupar authors\n",
    "id_authors_dict = {i: authors[counter] if i != \"N/A\" else [\"N/A\"] for counter, i in enumerate(idlist)}        \n",
    "        \n",
    "pubmed_list = []\n",
    "authors_list = []\n",
    "for key, vals in id_authors_dict.items():\n",
    "    if vals:\n",
    "        for val in vals:\n",
    "            pubmed_list.append(key)\n",
    "            authors_list.append(val)\n",
    "\n",
    "new_list_authors= []\n",
    "seen = set()\n",
    "for item in authors_list:\n",
    "    if item not in seen:\n",
    "        seen.add(item)\n",
    "        new_list_authors.append(item)\n",
    "\n",
    "#agrupar affi\n",
    "single_affiliation_list = []\n",
    "for i in affiliation:\n",
    "    single_affiliation_list.extend(i)\n",
    "    \n",
    "\n",
    "id_affiliation_dict = {i: [single_affiliation_list[counter]] if i != \"N/A\" else [\"N/A\"] for counter, i in enumerate(idlist)}\n",
    "\n",
    "pubmed_affi_list = []\n",
    "affi_list = []\n",
    "for key, vals in id_affiliation_dict.items():\n",
    "    if vals:\n",
    "        for val in vals:\n",
    "            pubmed_affi_list.append(key)\n",
    "            affi_list.append(val)\n",
    "    else:\n",
    "        pubmed_affi_list.append(key)\n",
    "        affi_list.append([\"N/A\"])\n",
    "\n",
    "new_list_affi= []\n",
    "seen = set()\n",
    "for item in single_affiliation_list :\n",
    "    if item not in seen:\n",
    "        seen.add(item)\n",
    "        new_list_affi.append(item)\n",
    "        \n",
    "#FINAL para protein_id - PROT_ID (NECESSÁRIO)\n",
    "list_pro=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "for info in records:\n",
    "    for i in info.features:\n",
    "        if i.type == \"CDS\":\n",
    "            pro= str(i.qualifiers[\"protein_id\"])\n",
    "            list_pro.append(pro)\n",
    "div= ', '.join(list_pro)\n",
    "h= div.replace(\"['\",'')\n",
    "hh= h.replace(\"']\",'')\n",
    "ID_PROT= hh.split(', ')\n",
    "print (ID_PROT) # is used below in Uniprot\n",
    "\n",
    "#FINAL para location, product e translation  - get_CDS_info(Ids) = [id_protein, location, product, translation] - NECESSÁRIO\n",
    "def get_CDS_info(id_list):\n",
    "    database = 'nucleotide'\n",
    "    email= 'rodrigoce9@gmail.com'\n",
    "    Entrez.email = email\n",
    "    handle = Entrez.efetch(db=database, id=id_list, rettype=\"gb\") \n",
    "    records = list(SeqIO.parse(handle,\"gb\"))\n",
    "    handle.close()\n",
    "    cds_location_list = []\n",
    "    \n",
    "    several_location = []\n",
    "    record_types={}\n",
    "    for i in id_list:\n",
    "        handle2 = Entrez.efetch(db=database, id=i, rettype=\"gb\")\n",
    "        for record in SeqIO.parse(handle2, \"genbank\"):\n",
    "            for feature in record.features:\n",
    "                record_types[feature.type]=record_types.get(feature.type,0)+1\n",
    "                cds_nonlocation, nonlocation, nonproduct, nontranslation = [], [], [], []\n",
    "            if 'CDS' not in record_types:\n",
    "                cds_nonlocation.append(\"N/A_CDS\")\n",
    "                nonlocation.append(\"N/A\")\n",
    "                nonproduct.append(\"N/A\")\n",
    "                nontranslation.append(\"N/A\")\n",
    "                cds_nonlocation.append(nonlocation)\n",
    "                cds_nonlocation.append(nonproduct)\n",
    "                cds_nonlocation.append(nontranslation)\n",
    "                cds_location_list.append(cds_nonlocation)\n",
    "#                 cds_location_list.append(nonlocation)\n",
    "                \n",
    "    handle2.close()\n",
    "    for info in records:\n",
    "        for i in info.features:\n",
    "            list_product, list_translation = [],[]\n",
    "            cds_location = []\n",
    "            product, translation = '', ''\n",
    "            if i.type == \"CDS\":\n",
    "                i_d = str(i.qualifiers[\"protein_id\"])\n",
    "                product = str(i.qualifiers[\"product\"])\n",
    "                translation =  str(i.qualifiers[\"translation\"])\n",
    "                cds_location.append(i_d)\n",
    "                if isinstance(i.location, CompoundLocation):\n",
    "                    for sub_location in i.location.parts:\n",
    "                        several_location.append(\"[{} : {}]\".format(sub_location.start, sub_location.end))\n",
    "                    cds_location.append(several_location)\n",
    "                    cds_location.append(product)\n",
    "                    cds_location.append(translation)\n",
    "                else:\n",
    "                    cds_location.append(\"[{} : {}]\".format(i.location.start, i.location.end))\n",
    "                    cds_location.append(product)\n",
    "                    cds_location.append(translation)\n",
    "                cds_location_list.append(cds_location)      \n",
    "    handle2.close()                    \n",
    "    return cds_location_list\n",
    "print(get_CDS_info(Ids))\n",
    "\n",
    "#FINAL result_dic = {Ids, ID_PROT} - NECESSÁRIO\n",
    "result_dict = {}\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist=Ids\n",
    "# idlist= Ids.remove('U49845')\n",
    "for ids in idlist:\n",
    "    list_pro=[]\n",
    "    handle = Entrez.efetch(db=database, id=ids, rettype=\"gb\") \n",
    "    records = list(SeqIO.parse(handle,\"gb\"))\n",
    "    handle.close()\n",
    "    for info in records:\n",
    "        list_pro.append(info.id)\n",
    "        for i in info.features:\n",
    "            if i.type == \"CDS\":\n",
    "                pro= str(i.qualifiers[\"protein_id\"])\n",
    "                list_pro.append(pro)\n",
    "        if len(list_pro)==1: #if no protein_id was found\n",
    "            result_dict[info.id] = \"N/A_CDS\"\n",
    "        else:\n",
    "            list_pro = [x.replace(\"['\",'').replace(\"']\",'') for x in list_pro]\n",
    "            result_dict[info.id] = ', '.join(list_pro[1:])\n",
    "print(result_dict)\n",
    "#key: id_genebank; values: id_Uniprot\n",
    "\n",
    "#final, mas não correr (código necessário para o caso se querermos links não diretos para uniprot )\n",
    "def url_get_id_p(dic):\n",
    "    url_list_p = []\n",
    "    for key, value in dic.items():\n",
    "        if value != 'N/A_CDS':\n",
    "            value_list = value.split(',')\n",
    "            for val in value_list:\n",
    "                val = val.strip()\n",
    "                url_id_p = \"https://www.uniprot.org/uniprotkb?query={}\".format(val)\n",
    "                url_list_p.append(url_id_p)\n",
    "    return url_list_p\n",
    "url_ids_protein=url_get_id_p(result_dict)\n",
    "print(url_ids_protein) # but we don't need to save this links\n",
    "\n",
    "#FINAL (usado em baixo) - NECESSÁRIO\n",
    "def url_get_id_p(string):\n",
    "#     print(string)\n",
    "    url_list_p=[ ]\n",
    "    url_id_p= \"https://www.uniprot.org/uniprotkb/{}/entry\".format(str(string))\n",
    "    url_list_p.append(url_id_p)\n",
    "    return ''.join(url_list_p)\n",
    "print(url_get_id_p)\n",
    "\n",
    "\n",
    "#FINAL - NECESSÁRIO\n",
    "WEBSITE_API = \"https://rest.uniprot.org\"\n",
    "fields = [\"accession\",\"organism_name\",\"protein_name\",\"cc_subcellular_location\",\"cc_function\", \"sequence\"]\n",
    "def get_url(url, **kwargs):\n",
    "    response = requests.get(url, **kwargs);\n",
    "    if not response.ok:\n",
    "        print(response.text)\n",
    "        response.raise_for_status()\n",
    "        sys.exit()\n",
    "    return response\n",
    "\n",
    "def get_field_for_id(ID_PROT, field):\n",
    "    response = get_url(\"{}/uniprotkb/search?query={}&fields={}&size=1&format=tsv\".format(WEBSITE_API,ID_PROT,field))\n",
    "    return str(response.content)\n",
    "\n",
    "def get_list_uniprot(ID_PROT, result_dict):\n",
    "    results = {}\n",
    "    result = {}\n",
    "    for first_key, first_value in result_dict.items(): \n",
    "        result = {}\n",
    "        for field in fields:\n",
    "            if first_value != 'N/A_CDS':\n",
    "                result[field] = get_field_for_id(first_value, field)\n",
    "            else:\n",
    "                result[field] = 'N/A'\n",
    "            results[first_key] = result   \n",
    "    \n",
    "    uniprot_final_list=[]\n",
    "    for key, value in results.items():\n",
    "        uniprot_list=[]\n",
    "        for x in value.values():\n",
    "#             print(key)\n",
    "            entry = re.search(r'b\\'Entry\\\\n(.+?(?=\\\\n\\'))', x, re.DOTALL)\n",
    "            function = re.match( r'b\\'Function \\[CC\\]\\\\n.{9} (.+?(?=\\\\n\\'))', x, re.DOTALL )\n",
    "            location_exist = re.search( r'b\\'Subcellular location \\[CC\\]\\\\nSUBCELLULAR LOCATION: (.+?(?=\\\\n\\'))', x, re.DOTALL )\n",
    "            location_notexist = re.search( r'b\\'Subcellular location \\[CC\\]\\\\n\\\\n\\'', x, re.DOTALL )   \n",
    "            sequence = re.search(r'b\\'Sequence\\\\n(.+?(?=\\\\n\\'))', x, re.DOTALL)\n",
    "            n_a = re.search(r'(N/A)', x, re.DOTALL)\n",
    "            if entry:\n",
    "                ent=entry.group(1)\n",
    "                uniprot_list.append(entry.group(1))            \n",
    "            if function:\n",
    "                uniprot_list.append(function.group(1))\n",
    "#                 print(function.group(0))\n",
    "#             if not function:\n",
    "#                 print('na')\n",
    "            if location_exist:\n",
    "                uniprot_list.append(location_exist.group(1))\n",
    "            if location_notexist:\n",
    "                uniprot_list.append(\"N/A\")\n",
    "            if sequence:\n",
    "                uniprot_list.append(sequence.group(1))\n",
    "                uniprot_list.append(len(sequence.group(1)))\n",
    "                uniprot_list.append(url_get_id_p(ent))\n",
    "            if n_a:\n",
    "                uniprot_list.append(n_a.group(1))\n",
    "                \n",
    "        uniprot_final_list.append(uniprot_list)\n",
    "        \n",
    "    return uniprot_final_list\n",
    "print(get_list_uniprot(ID_PROT,result_dict)) #list of lists (ID_Uniprot, Function, Subcelular location, Protein seq, length_aa)\n",
    "\n",
    "#FINAL [idgenebank, protein_id, id do uniprot]  -  NECESSÁRIO\n",
    "def join_ids_CDS(dic, uniprotID):\n",
    "    join_list_all = []\n",
    "    for key, value in dic.items():\n",
    "        join_list = []\n",
    "        join_list.append(key)\n",
    "        join_list.append(value)\n",
    "        join_list_all.append(join_list) \n",
    "    for index, values in enumerate(uniprotID):\n",
    "        join_list_all[index].append(uniprotID[index][0])\n",
    "    return join_list_all\n",
    "join_ids_CDS(result_dict,get_list_uniprot(ID_PROT, result_dict))\n",
    "\n",
    "#final [idgenebank, protein_id, id_uniprot,location, product, translation] (NÃO SEI SE ESTÁ A FUNCAR DIREITO - assumir que não)\n",
    "def join_lists(list1, list2):\n",
    "    result = []\n",
    "    for item1 in list1:\n",
    "        print(item1)\n",
    "        for item2 in list2:\n",
    "            print(item2)\n",
    "            string_item2 = str(item2[0])\n",
    "            string_item2 = string_item2.replace(\"[\",\"\").replace(\"]\",\"\").replace(\"'\",\"\")\n",
    "            if item1[1] == string_item2 or item1[1] == string_item2:\n",
    "                result.append(item1 + item2[1:])\n",
    "    return result\n",
    "join_lists(join_ids_CDS(result_dict,get_list_uniprot(ID_PROT, result_dict)), get_CDS_info(Ids))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "40bbca2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Erro na escrita na base de dados: 1062 (23000): Duplicate entry '91-310' for key 'PRIMARY'\n",
      "Erro na escrita na base de dados: 1062 (23000): Duplicate entry '91-57' for key 'PRIMARY'\n"
     ]
    }
   ],
   "source": [
    "#Povoação \"History\"\n",
    "\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    sql= \"INSERT INTO History (search, Genes_number_input, Day)   VALUES (%s, %s, %s)\"\n",
    "    val=(query, numero_genes, data_e_hora_atuais)\n",
    "    Cursor.execute(sql,val)\n",
    "    \n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "    \n",
    "search_id =[]\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    sql= \"Select ID_search FROM History\"\n",
    "    Cursor.execute(sql)\n",
    "    for row in Cursor:\n",
    "        search_id.append(str(row)) \n",
    "        \n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "    \n",
    "div= ', '.join(search_id)\n",
    "h= div.replace(\"(\",'')\n",
    "hh= h.replace(\",)\",'')\n",
    "SEARCH_ID= hh.split(', ')\n",
    "Hist= SEARCH_ID[-1]\n",
    "\n",
    "#Povoação \"Gene\"\n",
    "\n",
    "#tab = (\"INSERT INTO Gene (ID_genebank) VALUES ('{}')\".format(value)) # %s overwrites the same value\n",
    "\n",
    "des = (description)\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    for index, value in enumerate(Ids):\n",
    "        #print(value)\n",
    "        #print(des[index])\n",
    "        sql= \"INSERT INTO Gene (ID_genebank, Description, Organism, length, Adenina, Citosina, Guanina, Timina, Link, ID_search) VALUES ( %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\"\n",
    "        val=(value, des[index], Organismos[index], tam[index], Adenina[index], Citosina[index], Guanina[index], Timina[index], link_genebank[index], Hist)\n",
    "#         print(type(des[index]))\n",
    "        Cursor.execute(sql,val)\n",
    "#         for values in des:\n",
    "#             Cursor.execute(f\"INSERT INTO Gene ( Description)  VALUES ('{values}')\")\n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "    \n",
    "#Povoar \"PubMed\"\n",
    "\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    for index, value in enumerate(new_list_):\n",
    "        #print(value)\n",
    "        #print(des[index])\n",
    "        sql= \"INSERT INTO PubMed (ID_PumMed, title, Doi_number) VALUES (%s, %s, %s)\"\n",
    "        val=(str(new_list_[index]), str(titles[index]), str(doi_list[index]) )\n",
    "#         print(type(des[index]))\n",
    "        Cursor.execute(sql,val)\n",
    "#         for values in des:\n",
    "#             Cursor.execute(f\"INSERT INTO Gene ( Description)  VALUES ('{values}')\")\n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "    \n",
    "#retirar os valores AI de pubmed\n",
    "ID_AI=[]\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    sql= \"Select ID_AI_PubMed FROM PubMed\"\n",
    "    Cursor.execute(sql)\n",
    "    for row in Cursor:\n",
    "        #print(row)\n",
    "        ID_AI.append(str(row))       \n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "    \n",
    "   \n",
    "div_= ', '.join(ID_AI)\n",
    "h_= div_.replace(\"(\",'')\n",
    "hh_= h_.replace(\",)\",'')\n",
    "SEARCH_ID_= hh_.split(', ')\n",
    "#print(SEARCH_ID_)\n",
    "\n",
    "#Povação \"Gene-Pubmed\"\n",
    "number_map = {}\n",
    "next_number = int(SEARCH_ID_[0])\n",
    "relation = []\n",
    "for number in ID_PUB:\n",
    "    if number not in number_map:\n",
    "        number_map[number] = next_number\n",
    "        next_number += 1\n",
    "    relation.append(number_map[number])\n",
    "    \n",
    "\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    for index, value in enumerate(id_ncbii):\n",
    "        \n",
    "        sql= \"INSERT INTO gene_PubMed (ID_genebank, ID_AI_PubMed) VALUES (%s, %s)\"\n",
    "        val=(id_ncbii[index], relation[index])\n",
    "    \n",
    "        Cursor.execute(sql,val)\n",
    "\n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "    \n",
    "#Povação \"authors\"\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    for index, value in enumerate(new_list_authors):\n",
    "        \n",
    "        sql= \"INSERT INTO Authors (Name) VALUES (%s)\"\n",
    "        val=(new_list_authors[index],)\n",
    "    \n",
    "        Cursor.execute(sql,val)\n",
    "\n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "\n",
    "#retirar os valores AI de authors\n",
    "ID_AI_Authors=[]\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    sql= \"Select ID_Authors FROM Authors\"\n",
    "    Cursor.execute(sql)\n",
    "    for row in Cursor:\n",
    "        #print(row)\n",
    "        ID_AI_Authors.append(str(row))       \n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "    \n",
    "div_Authors= ', '.join(ID_AI_Authors)\n",
    "h_Authors= div_Authors.replace(\"(\",'')\n",
    "hh_Authors= h_Authors.replace(\",)\",'')\n",
    "SEARCH_ID_Authors= hh_Authors.split(', ')\n",
    "#print(SEARCH_ID_Authors)\n",
    "\n",
    "#Povação \"Pubmed-Authors\"\n",
    "number_map_ = {}\n",
    "next_number = int(SEARCH_ID_Authors[0])\n",
    "output_authors = []\n",
    "for number in authors_list:\n",
    "    if number not in number_map_:\n",
    "        number_map_[number] = next_number\n",
    "        next_number += 1\n",
    "    output_authors.append(number_map_[number])\n",
    "\n",
    "number_map_pub = {}\n",
    "next_number = int(SEARCH_ID_[0])\n",
    "output_pub = []\n",
    "for number in pubmed_list:\n",
    "    if number not in number_map_pub:\n",
    "        number_map_pub[number] = next_number\n",
    "        next_number += 1\n",
    "    output_pub.append(number_map_pub[number])\n",
    "\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    for index, value in enumerate(output_pub):\n",
    "        \n",
    "        sql= \"INSERT INTO PubMed_Authors (ID_AI_PubMed, ID_Authors) VALUES (%s, %s)\"\n",
    "        val=(str(output_pub[index]),str(output_authors[index]))\n",
    "    \n",
    "        Cursor.execute(sql,val)\n",
    "\n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "    \n",
    "#Povação \"affi\"\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    for index, value in enumerate(new_list_affi):\n",
    "        \n",
    "        sql= \"INSERT INTO Affiliation (Info) VALUES (%s)\"\n",
    "        val=(new_list_affi[index],)\n",
    "    \n",
    "        Cursor.execute(sql,val)\n",
    "\n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "\n",
    "    \n",
    "#retirar os valores AI de affi\n",
    "ID_AI_Affiliation=[]\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    sql= \"Select ID_Affiliation FROM Affiliation\"\n",
    "    Cursor.execute(sql)\n",
    "    for row in Cursor:\n",
    "        #print(row)\n",
    "        ID_AI_Affiliation.append(str(row))       \n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "    \n",
    "div_Affiliation= ', '.join(ID_AI_Affiliation)\n",
    "h_Affiliation= div_Affiliation.replace(\"(\",'')\n",
    "hh_Affiliation= h_Affiliation.replace(\",)\",'')\n",
    "SEARCH_ID_Affiliation= hh_Affiliation.split(', ')\n",
    "#print(SEARCH_ID_Affiliation)\n",
    "\n",
    "#Povação \"Pubmed-Affiliation\"\n",
    "number_map_affi_pub = {}\n",
    "next_number = int(SEARCH_ID_[0])\n",
    "output_pub_affi = []\n",
    "for number in pubmed_affi_list:\n",
    "    if number not in number_map_affi_pub:\n",
    "        number_map_affi_pub[number] = next_number\n",
    "        next_number += 1\n",
    "    output_pub_affi.append(number_map_affi_pub[number])\n",
    "    \n",
    "number_map_affi = {}\n",
    "next_number = int(SEARCH_ID_Affiliation[0])\n",
    "output_affi = []\n",
    "for number in affi_list:\n",
    "    if number not in number_map_affi:\n",
    "        number_map_affi[number] = next_number\n",
    "        next_number += 1\n",
    "    output_affi.append(number_map_affi[number])\n",
    "\n",
    "\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    for index, value in enumerate(output_pub_affi):\n",
    "        \n",
    "        sql= \"INSERT INTO PubMed_Affiliation (ID_AI_PubMed, ID_Affiliation) VALUES (%s, %s)\"\n",
    "        val=(str(output_pub_affi[index]),str(output_affi[index]))\n",
    "    \n",
    "        Cursor.execute(sql,val)\n",
    "\n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9c9197f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Erro na escrita na base de dados: 1452 (23000): Cannot add or update a child row: a foreign key constraint fails (`AP_db_KRG`.`CDS`, CONSTRAINT `uniprot` FOREIGN KEY (`ID_Uniprot`) REFERENCES `Uniprot` (`ID_Uniprot`) ON DELETE NO ACTION ON UPDATE NO ACTION)\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "get_list_uniprot() missing 1 required positional argument: 'result_dict'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[1;32mIn [32]\u001b[0m, in \u001b[0;36m<cell line: 48>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     47\u001b[0m Cursor \u001b[38;5;241m=\u001b[39m DataBase\u001b[38;5;241m.\u001b[39mcursor()\n\u001b[0;32m     48\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m     49\u001b[0m \u001b[38;5;66;03m#     sql = \"ALTER TABLE Uniprot AUTO_INCREMENT = 0\"\u001b[39;00m\n\u001b[0;32m     50\u001b[0m \u001b[38;5;66;03m#     Cursor.execute(sql)\u001b[39;00m\n\u001b[1;32m---> 51\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m index, value \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(\u001b[43mget_list_uniprot\u001b[49m\u001b[43m(\u001b[49m\u001b[43mID_PROT\u001b[49m\u001b[43m)\u001b[49m):\n\u001b[0;32m     52\u001b[0m         \u001b[38;5;66;03m#print(value)\u001b[39;00m\n\u001b[0;32m     53\u001b[0m         sql\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mINSERT INTO Uniprot (ID_Uniprot, Subcelular_Location, Function, Protein_sequence, length_aa, Link_Uniprot) VALUES (\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m, \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m, \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m, \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m, \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m, \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m)\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m     54\u001b[0m         val\u001b[38;5;241m=\u001b[39m(get_list_uniprot(ID_PROT)[index][\u001b[38;5;241m0\u001b[39m],get_list_uniprot(ID_PROT)[index][\u001b[38;5;241m1\u001b[39m], get_list_uniprot(ID_PROT)[index][\u001b[38;5;241m2\u001b[39m], get_list_uniprot(ID_PROT)[index][\u001b[38;5;241m3\u001b[39m], get_list_uniprot(ID_PROT)[index][\u001b[38;5;241m4\u001b[39m], get_list_uniprot(ID_PROT)[index][\u001b[38;5;241m5\u001b[39m])\n",
      "\u001b[1;31mTypeError\u001b[0m: get_list_uniprot() missing 1 required positional argument: 'result_dict'"
     ]
    }
   ],
   "source": [
    "#Povoação \"CDS\"\n",
    "\n",
    "import mysql.connector as SQLC\n",
    "import mysql.connector\n",
    "\n",
    "\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    for index, value in enumerate(Ids):\n",
    "        #print(value)\n",
    "        #print(des[index])\n",
    "        sql= \"INSERT INTO CDS (ID_CDS, Product, Translation, Location, ID_genebank, ID_Uniprot) VALUES (%s, %s, %s, %s, %s, %s)\"\n",
    "        val=(get_CDS_info(Ids)[index][0], get_CDS_info(Ids)[index][2], get_CDS_info(Ids)[index][3],get_CDS_info(Ids)[index][1],join_ids_CDS(result_dict,get_list_uniprot(ID_PROT, result_dict))[index][0],join_ids_CDS(result_dict,get_list_uniprot(ID_PROT, result_dict))[index][2])\n",
    "#         print(type(des[index]))\n",
    "        Cursor.execute(sql,val)\n",
    "#         for values in des:\n",
    "#             Cursor.execute(f\"INSERT INTO Gene ( Description)  VALUES ('{values}')\")\n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "    \n",
    "\n",
    "#Povoação \"Uniprot\"\n",
    "# print(uniprot_final_list) #(ID_Uniprot, Subcelular location, Function, Protein seq, length_aa)\n",
    "\n",
    "import mysql.connector as SQLC\n",
    "import mysql.connector\n",
    "\n",
    "\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "#     sql = \"ALTER TABLE Uniprot AUTO_INCREMENT = 0\"\n",
    "#     Cursor.execute(sql)\n",
    "    for index, value in enumerate(get_list_uniprot(ID_PROT)):\n",
    "        #print(value)\n",
    "        sql= \"INSERT INTO Uniprot (ID_Uniprot, Subcelular_Location, Function, Protein_sequence, length_aa, Link_Uniprot) VALUES (%s, %s, %s, %s, %s, %s)\"\n",
    "        val=(get_list_uniprot(ID_PROT)[index][0],get_list_uniprot(ID_PROT)[index][1], get_list_uniprot(ID_PROT)[index][2], get_list_uniprot(ID_PROT)[index][3], get_list_uniprot(ID_PROT)[index][4], get_list_uniprot(ID_PROT)[index][5])\n",
    "        Cursor.execute(sql,val)\n",
    "    \n",
    "    ID_AI=[]\n",
    "    sql= \"Select ID_AI FROM Uniprot\"\n",
    "    Cursor.execute(sql)\n",
    "    for row in Cursor:\n",
    "        ID_AI.append(str(row))\n",
    "    div_= ', '.join(ID_AI)\n",
    "    h_= div_.replace(\"(\",'')\n",
    "    hh_= h_.replace(\",)\",'')\n",
    "    SEARCH_ID_AI= hh_.split(', ')\n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "\n",
    "#para buscar os ids da tabela Uniprot, que sao AI\n",
    "ID_AI=[]\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "    sql= \"Select ID_AI FROM Uniprot\"\n",
    "    Cursor.execute(sql)\n",
    "    for row in Cursor:\n",
    "        #print(row)\n",
    "        ID_AI.append(str(row))       \n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()\n",
    "    \n",
    "div_= ', '.join(ID_AI)\n",
    "h_= div_.replace(\"(\",'')\n",
    "hh_= h_.replace(\",)\",'')\n",
    "SEARCH_ID_AI= hh_.split(', ')\n",
    "print(SEARCH_ID_AI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "bba5bfa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mysql.connector as SQLC\n",
    "import mysql.connector\n",
    "\n",
    "des = (description)\n",
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True # allows the change be done\n",
    "\n",
    "Cursor = DataBase.cursor()\n",
    "try:\n",
    "        \n",
    "        a= \"delete from PubMed_Affiliation\"\n",
    "        Z= \"delete from PubMed_Authors\"\n",
    "        b= \"delete from Affiliation\"\n",
    "        c= \"delete from Authors\"\n",
    "        d= \"delete from gene_PubMed\"\n",
    "        e= \"delete from PubMed\"\n",
    "        f= \"delete from CDS\"\n",
    "        g= \"delete from Uniprot\"\n",
    "        h= \"delete from Gene\"\n",
    "        i= \"delete from History\"     \n",
    "        \n",
    "        Cursor.execute(a)\n",
    "        Cursor.execute(Z)\n",
    "        Cursor.execute(b)\n",
    "        Cursor.execute(c)\n",
    "        Cursor.execute(d)\n",
    "        Cursor.execute(e)\n",
    "        Cursor.execute(f)\n",
    "        Cursor.execute(g)\n",
    "        Cursor.execute(h)\n",
    "        Cursor.execute(i)\n",
    "        \n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    DataBase.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8065986a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
