{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML \n",
    "from Bio.Blast import NCBIWWW\n",
    "import requests, sys, json\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from Bio import Medline\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from Bio import SearchIO\n",
    "from Bio.SwissProt import KeyWList\n",
    "from Bio import SwissProt\n",
    "from Bio import ExPASy\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Seq import Seq\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator\n",
    "from Bio.Phylo.TreeConstruction import DistanceTreeConstructor\n",
    "from Bio import Phylo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### web scraping (escolher condição e nº de genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "escolha o que quer pesquisar: irs\n",
      "escolha o nº de genes que quer obter (máximo 20): 5\n",
      "['8223391', '64880', '3376', '45785', '408438']\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "query= input('escolha o que quer pesquisar: ')\n",
    "def url_get(i):\n",
    "    url_list= [ ]\n",
    "    url = \"https://www.ncbi.nlm.nih.gov/gene/?term={}\".format( query )\n",
    "    url_list.append(url)\n",
    "    return url_list\n",
    "url_get(1)\n",
    "\n",
    "import requests, sys, json\n",
    "content = []\n",
    "for url in url_get(1):\n",
    "    r = requests.get(url)\n",
    "    content.append(r.content)\n",
    "#print(content)\n",
    "\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "\n",
    "# Parsing the HTML\n",
    "for c in content:\n",
    "    soup = BeautifulSoup(c, 'html.parser')\n",
    "    a= soup.get_text()\n",
    "    #print (a)\n",
    "\n",
    "import re\n",
    "existe = re.findall(r\"ID:\\s+\\d*(?=\\D)\", a, re.DOTALL)\n",
    "\n",
    "c= ', '.join(existe)\n",
    "h= c.replace('ID: ','')\n",
    "IDS= h.split(', ')\n",
    "\n",
    "n_genes= IDS[0:int(input('escolha o nº de genes que quer obter (máximo 20): '))]\n",
    "#print(n_genes)\n",
    "\n",
    "if n_genes == ['']:\n",
    "    print (\"Não encontramos resultado para a sua pesquisa\")\n",
    "else:\n",
    "    print (n_genes)\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tabela \"Gene\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar ids do ncbi:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AL329769.1\n",
      "X61083.1\n",
      "U02356.1\n"
     ]
    }
   ],
   "source": [
    "Ids=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "for info in records:\n",
    "    Ids.append(info.id)\n",
    "    print(info.id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar Links dos genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['https://www.ncbi.nlm.nih.gov/nuccore/AL329769.1',\n",
       " 'https://www.ncbi.nlm.nih.gov/nuccore/X61083.1',\n",
       " 'https://www.ncbi.nlm.nih.gov/nuccore/U02356.1']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import time\n",
    "def url_get_id(i):\n",
    "    url_list= [ ]\n",
    "    for id in i:\n",
    "        url = \"https://www.ncbi.nlm.nih.gov/nuccore/{}\".format( id )\n",
    "        url_list.append(url)\n",
    "    return url_list\n",
    "url_get_id(Ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Tetraodon nigroviridis genome survey sequence T3 end of clone 012H02 o - Nucleotide - NCBI\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Warning:\n",
      "\tThe NCBI web site requires JavaScript to function. \n",
      "\tmore...\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "An official website of the United States government\n",
      "\n",
      "Here's how you know\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "The .gov means it's official.\n",
      "\n",
      "\t\t\t\t\t\t\tFederal government websites often end in .gov or .mil. Before\n",
      "\t\t\t\t\t\t\tsharing sensitive information, make sure you're on a federal\n",
      "\t\t\t\t\t\t\tgovernment site.\n",
      "\t\t\t\t\t\t\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "The site is secure.\n",
      "\n",
      "\t\t\t\t\t\t\tThe https:// ensures that you are connecting to the\n",
      "\t\t\t\t\t\t\tofficial website and that any information you provide is encrypted\n",
      "\t\t\t\t\t\t\tand transmitted securely.\n",
      "\t\t\t\t\t\t\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Log in\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Show account info\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Close\n",
      "Account\n",
      "\n",
      "\n",
      "\t\t\t\t\t\t\tLogged in as:\n",
      "username\n",
      "\n",
      "\n",
      "\n",
      "Dashboard\n",
      "Publications\n",
      "Account settings\n",
      "Log out\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Access keys\n",
      "NCBI Homepage\n",
      "MyNCBI Homepage\n",
      "Main Content\n",
      "Main Navigation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Nucleotide\n",
      "\n",
      "Search databaseAll DatabasesAssemblyBiocollectionsBioProjectBioSampleBooksClinVarConserved DomainsdbGaPdbVarGeneGenomeGEO DataSetsGEO ProfilesGTRHomoloGeneIdentical Protein GroupsMedGenMeSHNLM CatalogNucleotideOMIMPMCPopSetProteinProtein ClustersProtein Family ModelsPubChem BioAssayPubChem CompoundPubChem SubstancePubMedSNPSRAStructureTaxonomyToolKitToolKitAllToolKitBookghSearch termSearchAdvancedHelp\n",
      "\n",
      "\n",
      "\n",
      "Result Filters\n",
      "\n",
      "\n",
      "\n",
      "GenBankFormatSummaryGenBankGenBank (full)FASTAFASTA (text)GraphicsASN.1Revision HistoryAccession ListGI ListApplySend to:Choose DestinationFileClipboardCollectionsAnalysis ToolFormatSummaryGenBankGenBank (full)FASTAASN.1XMLINSDSeq XMLTinySeq XMLFeature TableAccession ListGI ListGFF3Show GICreate FileAdd to ClipboardAdd to CollectionsChoose Sequence Analysis ToolBLASTPrimerBLASTSubmit\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Tetraodon nigroviridis genome survey sequence T3 end of clone 012H02 of library C from Tetraodon nigroviridis, genomic survey sequence\n",
      "GenBank: AL329769.1\n",
      "\n",
      "\n",
      "FASTA\n",
      "Graphics\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "no features\n",
      "\n",
      "Feature\n",
      "\n",
      "First\n",
      "Previous\n",
      "\n",
      "Next\n",
      "Last\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Display: \n",
      "FASTA\n",
      "GenBank\n",
      "Help\n",
      "Close\n",
      "\n",
      "\n",
      "Details\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Supplemental Content\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Change region shown\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Whole sequence\n",
      "\n",
      "\n",
      "Selected region\n",
      "\n",
      "\n",
      " from: \n",
      "\n",
      " to: \n",
      "\n",
      "\n",
      "\n",
      "Update View\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Customize view\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Basic Features\n",
      "\n",
      "\n",
      " All features\n",
      "\n",
      "\n",
      "\n",
      "                            Gene,\n",
      "                            RNA,\n",
      "                         and CDS features only\n",
      "\n",
      "\n",
      "\n",
      "Display options\n",
      "\n",
      "\n",
      " Show reverse complement\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " Show gap features\n",
      "\n",
      "\n",
      "\n",
      "Update View\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Analyze this sequence\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Run BLAST\n",
      "Find regions of similarity between this sequence and other sequences\n",
      "                    using BLAST.\n",
      "\n",
      "\n",
      "Pick Primers\n",
      "Design and test primers for this sequence using\n",
      "                            Primer-BLAST.\n",
      "\n",
      "\n",
      "Find in this Sequence\n",
      "Finds sub-sequences or patterns in the sequence and highlights the\n",
      "                        matching regions. The tool works with standard single letter nucleotide or\n",
      "                        protein codes including ambiguities and can match Prosite patterns in\n",
      "                        protein sequences. More...\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Related information\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "PubMed\n",
      "PubMed articles cited by Nucleotide sequence record\n",
      "\n",
      "\n",
      "Taxonomy\n",
      "Taxonomy sequences associated with Nucleotide record\n",
      "\n",
      "\n",
      "Full text in PMC\n",
      "Free full text articles in PMC\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Recent activity\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                Clear\n",
      "                            \n",
      "\n",
      "                                Turn Off                        \n",
      "                            \n",
      "\n",
      "                                Turn On                        \n",
      "                            \n",
      "\n",
      "\n",
      "\n",
      "Tetraodon nigroviridis genome survey sequence T3 end of clone 012H02 of library ...\n",
      "Tetraodon nigroviridis genome survey sequence T3 end of clone 012H02 of library C from Tetraodon nigroviridis, genomic survey sequencegi|8223391|emb|AL329769.1|\n",
      "Nucleotide\n",
      "\n",
      "\n",
      "Your browsing activity is empty.\n",
      "Activity recording is turned off.\n",
      "\n",
      "Turn recording back on\n",
      "\n",
      "See more...\n",
      "\n",
      "\n",
      "\n",
      "ClearFindFirstPreviousNextLastHelpClose\n",
      "                Finds sub-sequence or patterns in the sequence and highlights the matching region.\n",
      "                The tool works with standard single letter nucleotide \n",
      "                or protein codes including ambiguities and can match \n",
      "                Prosite patterns in protein sequences.\n",
      "                More...\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Follow NCBI\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Twitter\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Facebook\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "LinkedIn\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "GitHub\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "NCBI Insights Blog\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Connect with NLM\n",
      "\n",
      "\n",
      "\n",
      "Twitter\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Facebook\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Youtube\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "National Library of Medicine\n",
      "8600 Rockville Pike\n",
      "\t\t\t\t\t\t\tBethesda, MD 20894\n",
      "\n",
      "\n",
      "Web Policies\n",
      "FOIA\n",
      "HHS Vulnerability Disclosure\n",
      "\n",
      "\n",
      "Help\n",
      "Accessibility\n",
      "Careers\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "NLM\n",
      "\n",
      "\n",
      "NIH\n",
      "\n",
      "\n",
      "HHS\n",
      "\n",
      "\n",
      "USA.gov\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "R.meliloti plasmid derived nodL gene for prabable acetyltransferase - Nucleotide - NCBI\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Warning:\n",
      "\tThe NCBI web site requires JavaScript to function. \n",
      "\tmore...\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "An official website of the United States government\n",
      "\n",
      "Here's how you know\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "The .gov means it's official.\n",
      "\n",
      "\t\t\t\t\t\t\tFederal government websites often end in .gov or .mil. Before\n",
      "\t\t\t\t\t\t\tsharing sensitive information, make sure you're on a federal\n",
      "\t\t\t\t\t\t\tgovernment site.\n",
      "\t\t\t\t\t\t\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "The site is secure.\n",
      "\n",
      "\t\t\t\t\t\t\tThe https:// ensures that you are connecting to the\n",
      "\t\t\t\t\t\t\tofficial website and that any information you provide is encrypted\n",
      "\t\t\t\t\t\t\tand transmitted securely.\n",
      "\t\t\t\t\t\t\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Log in\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Show account info\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Close\n",
      "Account\n",
      "\n",
      "\n",
      "\t\t\t\t\t\t\tLogged in as:\n",
      "username\n",
      "\n",
      "\n",
      "\n",
      "Dashboard\n",
      "Publications\n",
      "Account settings\n",
      "Log out\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Access keys\n",
      "NCBI Homepage\n",
      "MyNCBI Homepage\n",
      "Main Content\n",
      "Main Navigation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Nucleotide\n",
      "\n",
      "Search databaseAll DatabasesAssemblyBiocollectionsBioProjectBioSampleBooksClinVarConserved DomainsdbGaPdbVarGeneGenomeGEO DataSetsGEO ProfilesGTRHomoloGeneIdentical Protein GroupsMedGenMeSHNLM CatalogNucleotideOMIMPMCPopSetProteinProtein ClustersProtein Family ModelsPubChem BioAssayPubChem CompoundPubChem SubstancePubMedSNPSRAStructureTaxonomyToolKitToolKitAllToolKitBookghSearch termSearchAdvancedHelp\n",
      "\n",
      "\n",
      "\n",
      "Result Filters\n",
      "\n",
      "\n",
      "\n",
      "GenBankFormatSummaryGenBankGenBank (full)FASTAFASTA (text)GraphicsASN.1Revision HistoryAccession ListGI ListApplySend to:Complete RecordCoding SequencesGene FeaturesChoose DestinationFileClipboardCollectionsAnalysis ToolFormatSummaryGenBankGenBank (full)FASTAASN.1XMLINSDSeq XMLTinySeq XMLFeature TableAccession ListGI ListGFF3Show GICreate FileAdd to ClipboardAdd to CollectionsChoose Sequence Analysis ToolBLASTPrimerBLASTSubmitDownload features.FormatFASTA NucleotideFASTA ProteinCreate\n",
      "                FileDownload gene features.FormatFASTA NucleotideCreate File\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "R.meliloti plasmid derived nodL gene for prabable acetyltransferase\n",
      "GenBank: X61083.1\n",
      "\n",
      "\n",
      "FASTA\n",
      "Graphics\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "no features\n",
      "\n",
      "Feature\n",
      "\n",
      "First\n",
      "Previous\n",
      "\n",
      "Next\n",
      "Last\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Display: \n",
      "FASTA\n",
      "GenBank\n",
      "Help\n",
      "Close\n",
      "\n",
      "\n",
      "Details\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Supplemental Content\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Change region shown\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Whole sequence\n",
      "\n",
      "\n",
      "Selected region\n",
      "\n",
      "\n",
      " from: \n",
      "\n",
      " to: \n",
      "\n",
      "\n",
      "\n",
      "Update View\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Customize view\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Basic Features\n",
      "\n",
      "\n",
      " All features\n",
      "\n",
      "\n",
      "\n",
      "                            Gene,\n",
      "                            RNA,\n",
      "                         and CDS features only\n",
      "\n",
      "\n",
      "\n",
      "Display options\n",
      "\n",
      "\n",
      " Show reverse complement\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " Show gap features\n",
      "\n",
      "\n",
      "\n",
      "Update View\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Analyze this sequence\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Run BLAST\n",
      "Find regions of similarity between this sequence and other sequences\n",
      "                    using BLAST.\n",
      "\n",
      "\n",
      "Pick Primers\n",
      "Design and test primers for this sequence using\n",
      "                            Primer-BLAST.\n",
      "\n",
      "\n",
      "Highlight Sequence Features\n",
      "Opens the Highlight Feature Bar and highlights feature annotations\n",
      "                        from the FEATURES table of the record. The Highlight Feature Bar can be used\n",
      "                        to navigate to and highlight other features and provides links to display\n",
      "                        the highlighted region separately. Links in the FEATURES table will also\n",
      "                        highlight the corresponding region of the sequence. More...\n",
      "\n",
      "\n",
      "Find in this Sequence\n",
      "Finds sub-sequences or patterns in the sequence and highlights the\n",
      "                        matching regions. The tool works with standard single letter nucleotide or\n",
      "                        protein codes including ambiguities and can match Prosite patterns in\n",
      "                        protein sequences. More...\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Related information\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Protein\n",
      "Protein sequence from Nucleotide coding region\n",
      "\n",
      "\n",
      "PubMed\n",
      "PubMed articles cited by Nucleotide sequence record\n",
      "\n",
      "\n",
      "Taxonomy\n",
      "Taxonomy sequences associated with Nucleotide record\n",
      "\n",
      "\n",
      "Full text in PMC\n",
      "Free full text articles in PMC\n",
      "\n",
      "\n",
      "PubMed (Weighted)\n",
      "Links to pubmed\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Recent activity\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                Clear\n",
      "                            \n",
      "\n",
      "                                Turn Off                        \n",
      "                            \n",
      "\n",
      "                                Turn On                        \n",
      "                            \n",
      "\n",
      "\n",
      "\n",
      "R.meliloti plasmid derived nodL gene for prabable acetyltransferase\n",
      "R.meliloti plasmid derived nodL gene for prabable acetyltransferasegi|45785|emb|X61083.1|\n",
      "Nucleotide\n",
      "\n",
      "\n",
      "Your browsing activity is empty.\n",
      "Activity recording is turned off.\n",
      "\n",
      "Turn recording back on\n",
      "\n",
      "See more...\n",
      "\n",
      "\n",
      "\n",
      "ClearFindFirstPreviousNextLastHelpClose\n",
      "                Finds sub-sequence or patterns in the sequence and highlights the matching region.\n",
      "                The tool works with standard single letter nucleotide \n",
      "                or protein codes including ambiguities and can match \n",
      "                Prosite patterns in protein sequences.\n",
      "                More...\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Follow NCBI\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Twitter\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Facebook\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "LinkedIn\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "GitHub\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "NCBI Insights Blog\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Connect with NLM\n",
      "\n",
      "\n",
      "\n",
      "Twitter\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Facebook\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Youtube\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "National Library of Medicine\n",
      "8600 Rockville Pike\n",
      "\t\t\t\t\t\t\tBethesda, MD 20894\n",
      "\n",
      "\n",
      "Web Policies\n",
      "FOIA\n",
      "HHS Vulnerability Disclosure\n",
      "\n",
      "\n",
      "Help\n",
      "Accessibility\n",
      "Careers\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "NLM\n",
      "\n",
      "\n",
      "NIH\n",
      "\n",
      "\n",
      "HHS\n",
      "\n",
      "\n",
      "USA.gov\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Profundulus guatemalensis tyrosine kinase (X-src) gene, partial cds - Nucleotide - NCBI\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Warning:\n",
      "\tThe NCBI web site requires JavaScript to function. \n",
      "\tmore...\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "An official website of the United States government\n",
      "\n",
      "Here's how you know\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "The .gov means it's official.\n",
      "\n",
      "\t\t\t\t\t\t\tFederal government websites often end in .gov or .mil. Before\n",
      "\t\t\t\t\t\t\tsharing sensitive information, make sure you're on a federal\n",
      "\t\t\t\t\t\t\tgovernment site.\n",
      "\t\t\t\t\t\t\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "The site is secure.\n",
      "\n",
      "\t\t\t\t\t\t\tThe https:// ensures that you are connecting to the\n",
      "\t\t\t\t\t\t\tofficial website and that any information you provide is encrypted\n",
      "\t\t\t\t\t\t\tand transmitted securely.\n",
      "\t\t\t\t\t\t\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Log in\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Show account info\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Close\n",
      "Account\n",
      "\n",
      "\n",
      "\t\t\t\t\t\t\tLogged in as:\n",
      "username\n",
      "\n",
      "\n",
      "\n",
      "Dashboard\n",
      "Publications\n",
      "Account settings\n",
      "Log out\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Access keys\n",
      "NCBI Homepage\n",
      "MyNCBI Homepage\n",
      "Main Content\n",
      "Main Navigation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Nucleotide\n",
      "\n",
      "Search databaseAll DatabasesAssemblyBiocollectionsBioProjectBioSampleBooksClinVarConserved DomainsdbGaPdbVarGeneGenomeGEO DataSetsGEO ProfilesGTRHomoloGeneIdentical Protein GroupsMedGenMeSHNLM CatalogNucleotideOMIMPMCPopSetProteinProtein ClustersProtein Family ModelsPubChem BioAssayPubChem CompoundPubChem SubstancePubMedSNPSRAStructureTaxonomyToolKitToolKitAllToolKitBookghSearch termSearchAdvancedHelp\n",
      "\n",
      "\n",
      "\n",
      "Result Filters\n",
      "\n",
      "\n",
      "\n",
      "GenBankFormatSummaryGenBankGenBank (full)FASTAFASTA (text)GraphicsASN.1Revision HistoryAccession ListGI ListApplySend to:Complete RecordCoding SequencesGene FeaturesChoose DestinationFileClipboardCollectionsAnalysis ToolFormatSummaryGenBankGenBank (full)FASTAASN.1XMLINSDSeq XMLTinySeq XMLFeature TableAccession ListGI ListGFF3Show GICreate FileAdd to ClipboardAdd to CollectionsChoose Sequence Analysis ToolBLASTPrimerBLASTSubmitDownload features.FormatFASTA NucleotideFASTA ProteinCreate\n",
      "                FileDownload gene features.FormatFASTA NucleotideCreate File\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Profundulus guatemalensis tyrosine kinase (X-src) gene, partial cds\n",
      "GenBank: U02356.1\n",
      "\n",
      "\n",
      "FASTA\n",
      "Graphics\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "no features\n",
      "\n",
      "Feature\n",
      "\n",
      "First\n",
      "Previous\n",
      "\n",
      "Next\n",
      "Last\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Display: \n",
      "FASTA\n",
      "GenBank\n",
      "Help\n",
      "Close\n",
      "\n",
      "\n",
      "Details\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Supplemental Content\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Change region shown\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Whole sequence\n",
      "\n",
      "\n",
      "Selected region\n",
      "\n",
      "\n",
      " from: \n",
      "\n",
      " to: \n",
      "\n",
      "\n",
      "\n",
      "Update View\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Customize view\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Basic Features\n",
      "\n",
      "\n",
      " All features\n",
      "\n",
      "\n",
      "\n",
      "                            Gene,\n",
      "                            RNA,\n",
      "                         and CDS features only\n",
      "\n",
      "\n",
      "\n",
      "Display options\n",
      "\n",
      "\n",
      " Show reverse complement\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " Show gap features\n",
      "\n",
      "\n",
      "\n",
      "Update View\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Analyze this sequence\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Run BLAST\n",
      "Find regions of similarity between this sequence and other sequences\n",
      "                    using BLAST.\n",
      "\n",
      "\n",
      "Pick Primers\n",
      "Design and test primers for this sequence using\n",
      "                            Primer-BLAST.\n",
      "\n",
      "\n",
      "Highlight Sequence Features\n",
      "Opens the Highlight Feature Bar and highlights feature annotations\n",
      "                        from the FEATURES table of the record. The Highlight Feature Bar can be used\n",
      "                        to navigate to and highlight other features and provides links to display\n",
      "                        the highlighted region separately. Links in the FEATURES table will also\n",
      "                        highlight the corresponding region of the sequence. More...\n",
      "\n",
      "\n",
      "Find in this Sequence\n",
      "Finds sub-sequences or patterns in the sequence and highlights the\n",
      "                        matching regions. The tool works with standard single letter nucleotide or\n",
      "                        protein codes including ambiguities and can match Prosite patterns in\n",
      "                        protein sequences. More...\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Related information\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Protein\n",
      "Protein sequence from Nucleotide coding region\n",
      "\n",
      "\n",
      "PubMed\n",
      "PubMed articles cited by Nucleotide sequence record\n",
      "\n",
      "\n",
      "Taxonomy\n",
      "Taxonomy sequences associated with Nucleotide record\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Recent activity\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                Clear\n",
      "                            \n",
      "\n",
      "                                Turn Off                        \n",
      "                            \n",
      "\n",
      "                                Turn On                        \n",
      "                            \n",
      "\n",
      "\n",
      "\n",
      "Profundulus guatemalensis tyrosine kinase (X-src) gene, partial cds\n",
      "Profundulus guatemalensis tyrosine kinase (X-src) gene, partial cdsgi|408438|gb|U02356.1|PGXSRC\n",
      "Nucleotide\n",
      "\n",
      "\n",
      "Your browsing activity is empty.\n",
      "Activity recording is turned off.\n",
      "\n",
      "Turn recording back on\n",
      "\n",
      "See more...\n",
      "\n",
      "\n",
      "\n",
      "ClearFindFirstPreviousNextLastHelpClose\n",
      "                Finds sub-sequence or patterns in the sequence and highlights the matching region.\n",
      "                The tool works with standard single letter nucleotide \n",
      "                or protein codes including ambiguities and can match \n",
      "                Prosite patterns in protein sequences.\n",
      "                More...\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Follow NCBI\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Twitter\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Facebook\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "LinkedIn\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "GitHub\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "NCBI Insights Blog\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Connect with NLM\n",
      "\n",
      "\n",
      "\n",
      "Twitter\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Facebook\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Youtube\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "National Library of Medicine\n",
      "8600 Rockville Pike\n",
      "\t\t\t\t\t\t\tBethesda, MD 20894\n",
      "\n",
      "\n",
      "Web Policies\n",
      "FOIA\n",
      "HHS Vulnerability Disclosure\n",
      "\n",
      "\n",
      "Help\n",
      "Accessibility\n",
      "Careers\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "NLM\n",
      "\n",
      "\n",
      "NIH\n",
      "\n",
      "\n",
      "HHS\n",
      "\n",
      "\n",
      "USA.gov\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "content = []\n",
    "for url in url_get_id(Ids):\n",
    "    r = requests.get(url)\n",
    "    content.append(r.content)\n",
    "print(content)\n",
    "\n",
    "\n",
    "\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "\n",
    "# Parsing the HTML\n",
    "for c in content:\n",
    "    soup = BeautifulSoup(c, 'html.parser')\n",
    "\n",
    "    # Procurar um tag meta com um determinado atributo\n",
    "    lines = soup.find_all('meta', {'name':\"ncbi_uidlist\"} )\n",
    "\n",
    "    id = \"\"\n",
    "    url = \"\"\n",
    "    for line in lines:\n",
    "        #print(line)\n",
    "        #if 'name' in line.attrs:\n",
    "        #    print(line.attrs['name'])\n",
    "        if 'content' in line.attrs:\n",
    "            # print(line.attrs['content'])\t\t\n",
    "            id = line.attrs['content']\n",
    "\n",
    "    if id:\n",
    "        url = \"https://www.ncbi.nlm.nih.gov/sviewer/viewer.fcgi?id={}&db=nuccore&report=genbank&conwithfeat=on&hide-cdd=on&retmode=text&maxdownloadsize=5000000\".format(id)\n",
    "\n",
    "    r2 = requests.get(url )\n",
    "    r4= r2.content.decode('utf-8')\n",
    "    print (r4, '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar organismos: (novo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#organismos\n",
    "\n",
    "Organismos=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "for info in records:\n",
    "    Organismos.append(info.annotations['organism'])\n",
    "print(Organismos)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar seq dos genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Biopython\n",
    "Ids=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "seq=[]\n",
    "for info in records:\n",
    "    print(info.id)\n",
    "    print(info.seq[0:50],'...',info.seq[-10:])\n",
    "    print()\n",
    "    seq.append(f'{info.seq[0:50]}...{info.seq[-10:]}')\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar % de Alanina:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[25.674157303370787, 25.27881040892193, 25.89641434262948]\n"
     ]
    }
   ],
   "source": [
    "Alanina=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "seq=[]\n",
    "for info in records:\n",
    "    c=info.seq\n",
    "    count= c.count('A')\n",
    "    perc= (count/len(c)*100)\n",
    "    Alanina.append(perc)\n",
    "print (Alanina)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar % de Citosina:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Citosina=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "seq=[]\n",
    "for info in records:\n",
    "    c=info.seq\n",
    "    count= c.count('C')\n",
    "    perc= (count/len(c)*100)\n",
    "    Citosina.append(perc)\n",
    "print (Citosina)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar % de Guanina:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Guanina=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "seq=[]\n",
    "for info in records:\n",
    "    c=info.seq\n",
    "    count= c.count('G')\n",
    "    perc= (count/len(c)*100)\n",
    "    Guanina.append(perc)\n",
    "print (Guanina)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar % de Timina:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[27.191011235955052, 24.287484510532835, 19.721115537848604]\n"
     ]
    }
   ],
   "source": [
    "Timina=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "seq=[]\n",
    "for info in records:\n",
    "    c=info.seq\n",
    "    count= c.count('T')\n",
    "    perc= (count/len(c)*100)\n",
    "    Timina.append(perc)\n",
    "print (Timina)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar gene description:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AL329769.1\n",
      "Tetraodon nigroviridis genome survey sequence T3 end of clone 012H02 of library C from Tetraodon nigroviridis, genomic survey sequence\n",
      "X61083.1\n",
      "R.meliloti plasmid derived nodL gene for prabable acetyltransferase\n",
      "U02356.1\n",
      "Profundulus guatemalensis tyrosine kinase (X-src) gene, partial cds\n"
     ]
    }
   ],
   "source": [
    "Ids=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "description=[]\n",
    "for info in records:\n",
    "    print(info.id)\n",
    "    print(info.description)\n",
    "    description.append(info.description)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar Data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "for info in records:\n",
    "    data.append(info.annotations['date'])\n",
    "print(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar tamanho:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'n_genes' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [1]\u001b[0m, in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m database \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnucleotide\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      2\u001b[0m email\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrodrigoce9@gmail.com\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 3\u001b[0m idlist\u001b[38;5;241m=\u001b[39m \u001b[43mn_genes\u001b[49m\n\u001b[1;32m      4\u001b[0m handle \u001b[38;5;241m=\u001b[39m Entrez\u001b[38;5;241m.\u001b[39mefetch(db\u001b[38;5;241m=\u001b[39mdatabase, \u001b[38;5;28mid\u001b[39m\u001b[38;5;241m=\u001b[39midlist, rettype\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgb\u001b[39m\u001b[38;5;124m\"\u001b[39m) \n\u001b[1;32m      5\u001b[0m records \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(SeqIO\u001b[38;5;241m.\u001b[39mparse(handle,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgb\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n",
      "\u001b[0;31mNameError\u001b[0m: name 'n_genes' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "tam=[]\n",
    "for info in records:\n",
    "    tam.append(len(info.seq))\n",
    "\n",
    "print(tam)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tabela \"CDS\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar quantidade de cds e a sua localização:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[0;32mIn [69]\u001b[0m, in \u001b[0;36m<cell line: 12>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     11\u001b[0m teste\u001b[38;5;241m=\u001b[39m[]\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m teste \u001b[38;5;241m!=\u001b[39m \u001b[38;5;28mlen\u001b[39m(Ids):\n\u001b[0;32m---> 13\u001b[0m     \u001b[43mteste\u001b[49m\u001b[38;5;241m.\u001b[39mappend(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msem\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     14\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m info \u001b[38;5;129;01min\u001b[39;00m records:\n\u001b[1;32m     15\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(info\u001b[38;5;241m.\u001b[39mfeatures)):\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "Ids=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "featcds = [ ]\n",
    "position=0\n",
    "pos=[]\n",
    "for info in records:\n",
    "    for i in range(len(info.features)):\n",
    "        if info.features[i].type == \"CDS\":\n",
    "            cds= i\n",
    "            a= str(cds)\n",
    "            b= len(a)\n",
    "            print (b)\n",
    "            featcds.append(b)\n",
    "            #position=i\n",
    "            #pos.append(info.features[position].location)\n",
    "            #print (pos)\n",
    "    #for h in info.features:\n",
    "        #if h.type[\"CDS\"] == none:\n",
    "            #featcds.append(\"sem resultados\")\n",
    "print(teste)     \n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tabela \"Uniprot\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar Ids de proteinas:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_pro=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "for info in records:\n",
    "    for i in info.features:\n",
    "        if i.type == \"CDS\":\n",
    "            \n",
    "            pro= str(i.qualifiers[\"protein_id\"])\n",
    "            list_pro.append(pro)\n",
    "            \n",
    "div= ', '.join(list_pro)\n",
    "h= div.replace(\"['\",'')\n",
    "hh= h.replace(\"']\",'')\n",
    "ID_PROT= hh.split(', ')\n",
    "print (ID_PROT) # is used below in Uniprot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar Links de proteinas:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "ID_PROT\n",
    "def url_get_id_p(a):\n",
    "    url_list_p=[ ]\n",
    "    for id_p in ID_PROT:\n",
    "        url_id_p= \"https://www.uniprot.org/uniprotkb?query={}\".format(id_p)\n",
    "        url_list_p.append(url_id_p)\n",
    "    return url_list_p\n",
    "url_ids_protein=url_get_id_p(ID_PROT)\n",
    "print(url_ids_protein)  # but we don't need to save this links"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar id proteina (uniprot):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_url(url, **kwargs):\n",
    "    response = requests.get(url, **kwargs);\n",
    "\n",
    "    if not response.ok:\n",
    "        print(response.text)\n",
    "        response.raise_for_status()\n",
    "        sys.exit()\n",
    "\n",
    "    return response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#final version\n",
    "WEBSITE_API = \"https://rest.uniprot.org\"\n",
    "fields = [\"accession\",\"organism_name\",\"protein_name\",\"cc_subcellular_location\",\"cc_function\", \"sequence\"]\n",
    "\n",
    "def get_field_for_id(ID_PROT, field):\n",
    "    response = get_url(\"{}/uniprotkb/search?query={} AND (reviewed:true)&fields={}&size=1&format=tsv\".format(WEBSITE_API,ID_PROT,field))\n",
    "    return str(response.content)\n",
    "\n",
    "\n",
    "result = {}\n",
    "for field in fields:\n",
    "    result[field] = [get_field_for_id(i, field) for i in ID_PROT]\n",
    "\n",
    "for field, values in result.items():\n",
    "#     print(f\"Results for {field}\")\n",
    "#     print(values)\n",
    "    #regex\n",
    "    for x in values:\n",
    "        entry = re.search(r'b\\'Entry\\\\n(.+?(?=\\\\n\\'))', x, re.DOTALL)\n",
    "        organism = re.search(r'b\\'Organism\\\\n(.+?(?=\\\\n\\'))', x, re.DOTALL)\n",
    "        function = re.search( r'b\\'Function \\[(CC)\\]\\\\n.{9} (.+?(?=\\\\n\\'))', x, re.DOTALL )\n",
    "        location = re.search( r'b\\'Subcellular location .{27} (.+?(?=\\\\n\\'))', x, re.DOTALL )\n",
    "        sequence = re.search(r'b\\'Sequence\\\\n(.+?(?=\\\\n\\'))', x, re.DOTALL)\n",
    "        if entry:\n",
    "            e = re.match(r'b\\'Entry\\\\n(.+?(?=\\\\n\\'))', x, re.DOTALL )\n",
    "            print(e.group(1))\n",
    "        elif organism:\n",
    "            o = re.match(r'b\\'Organism\\\\n(.+?(?=\\\\n\\'))', x, re.DOTALL )\n",
    "            print(o.group(1))\n",
    "        elif function:\n",
    "            f = re.match(r'b\\'Function .{15} (.+?(?=\\\\n\\'))', x, re.DOTALL )\n",
    "            print(f.group(1))\n",
    "        elif location:\n",
    "            l = re.match(r'b\\'Subcellular location .{27} (.+?(?=\\\\n\\'))', x, re.DOTALL )\n",
    "            print(l.group(1))\n",
    "        elif sequence:\n",
    "            s = re.match(r'b\\'Sequence\\\\n(.+?(?=\\\\n\\'))', x, re.DOTALL)\n",
    "            print(s.group(1) , len(s.group(1)) )\n",
    "#         else:\n",
    "#             print(\"fail\")\n",
    "#ADD CONDITION FOR EMPY RESULT (MEANS THE ID PROTEIN IS NOT REVIEWED)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tabela \"Pubmed\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar id de artigos:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['https://pubmed.ncbi.nlm.nih.gov/?term=Tetraodon nigroviridis genome survey sequence T3 end of clone 012H02 of library C from Tetraodon nigroviridis, genomic survey sequence', 'https://pubmed.ncbi.nlm.nih.gov/?term=R.meliloti plasmid derived nodL gene for prabable acetyltransferase', 'https://pubmed.ncbi.nlm.nih.gov/?term=Profundulus guatemalensis tyrosine kinase (X-src) gene, partial cds']\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Buscar artigos:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ids=[]\n",
    "database = 'PubMed'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= n_genes\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"medline\", retmode=\"text\") \n",
    "records = Medline.parse(handle)\n",
    "description=[]\n",
    "for info in records:\n",
    "    print(\"title:\", info.get(\"TI\", \"-\"))\n",
    "    print('authors: ', info.get(\"AU\", \"-\"))\n",
    "    print(\"source:\", info.get(\"SO\", \"-\"))\n",
    "    print(\"affiliation: \", info.get(\"AD\", \"-\") ) # FAZER O IF/ELSE PARA O CASO DE NÃO TER AFFILIATION\n",
    "    print()"
   ]
  }
 ],
 "metadata": {
  "authors": [
   {
    "name": "Jorge Gustavo Rocha"
   }
  ],
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
